packages/SwingSet/test/test-comms.js

```diff
@@ -98,54 +120,67 @@ test('receive', t => {
   const { syscall, sends } = mockSyscall();
   const d = buildCommsDispatch(syscall, 'fakestate', 'fakehelpers');
   const { state, clistKit } = debugState.get(d);
-  const { provideRemoteForLocal, getRemoteForLocal } = clistKit;
+  const {
+    provideLocalForKernel,
+    provideRemoteForLocal,
+    getRemoteForLocal,
+  } = clistKit;
   // add the remote, and an object to send at
   const transmitterID = 'o-1';
-  const bob = 'o-10';
+  const bobKernel = 'o-5';
+  const bobLocal = provideLocalForKernel(bobKernel);
   const { remoteID, receiverID } = addRemote(state, 'remote1', transmitterID);
-  const remoteBob = flipRemoteSlot(provideRemoteForLocal(remoteID, bob));
-  t.is(remoteBob, 'ro+20');
+  const bobRemote = flipRemoteSlot(provideRemoteForLocal(remoteID, bobLocal));
+  t.is(bobRemote, 'ro+20');
 
   // now pretend the transport layer received a message from remote1, as if
   // the remote machine had performed bob!foo()
   d.deliver(
     receiverID,
     'receive',
-    encodeArgs(`deliver:${remoteBob}:foo:;argsbytes`),
+    encodeArgs(`deliver:${bobRemote}:foo:;argsbytes`),
     null,
   );
-  t.deepEqual(sends.shift(), [bob, 'foo', capdata('argsbytes')]);
+  t.deepEqual(sends.shift(), [bobKernel, 'foo', capdata('argsbytes')]);
 
   // bob!bar(alice, bob)
   d.deliver(
     receiverID,
     'receive',
-    encodeArgs(`deliver:${remoteBob}:bar::ro-20:${remoteBob};argsbytes`),
+    encodeArgs(`deliver:${bobRemote}:bar::ro-20:${bobRemote};argsbytes`),
     null,
   );
-  t.deepEqual(sends.shift(), [bob, 'bar', capdata('argsbytes', ['o+11', bob])]);
+  t.deepEqual(sends.shift(), [
+    bobKernel,
+    'bar',
+    capdata('argsbytes', ['o+31', bobKernel]),
+  ]);
   // if we were to send o+11, the other side should get ro+20, which is alice
-  t.is(getRemoteForLocal(remoteID, 'o+11'), 'ro+20');
+  t.is(getRemoteForLocal(remoteID, 'lo11'), 'ro+20');
 
   // bob!bar(alice, bob)
   d.deliver(
     receiverID,
     'receive',
-    encodeArgs(`deliver:${remoteBob}:bar::ro-20:${remoteBob};argsbytes`),
+    encodeArgs(`deliver:${bobRemote}:bar::ro-20:${bobRemote};argsbytes`),
     null,
   );
-  t.deepEqual(sends.shift(), [bob, 'bar', capdata('argsbytes', ['o+11', bob])]);
+  t.deepEqual(sends.shift(), [
+    bobKernel,
+    'bar',
+    capdata('argsbytes', ['o+31', bobKernel]),
+  ]);
 
   // bob!cat(alice, bob, ayana)
   d.deliver(
     receiverID,
     'receive',
-    encodeArgs(`deliver:${remoteBob}:cat::ro-20:${remoteBob}:ro-21;argsbytes`),
+    encodeArgs(`deliver:${bobRemote}:cat::ro-20:${bobRemote}:ro-21;argsbytes`),
     null,
   );
   t.deepEqual(sends.shift(), [
-    bob,
+    bobKernel,
     'cat',
-    capdata('argsbytes', ['o+11', bob, 'o+12']),
+    capdata('argsbytes', ['o+31', bobKernel, 'o+32']),
```

Comment:
I don't understand what you are asking for here.

Reply:
The `o+31` value is chosen by the comms vat (specifically the initial counter offset, and the number of allocations that have happened since comms was initialized). The test is somewhat fragile, in that someone who changes that counter offset might think they've broken something when the `o+47` (or whatever) emitted by the new comms code doesn't match the `o+31` that this test is expecting. But in fact it's the test being overspecified: we don't care that it emits `o+31` or `o+47` or anything in particular, but we *do* care that it's consistent with everything else it's emitted.

The first time `o+31` shows up, in the first `syscal.send(bobKernel, 'bar')` that we examine on line 153. From outside the comms vat, we know there are supposed to be three correlated identifiers (remote, local, kernel-facing) that all talk about the same object "alice". The remote one is `ro-20`, as delivered on line 150, which we just made up (since remote systems can make up any import ID they want). We don't know what the local ID is for sure, but the current allocation code happens to translate that into `lo11`. We don'also don't know what the kernel-facing ID is for sure, but it currently allocates `o+31`. When we examine the comms vat's `syscall.send` on line 153, and we expect to see `o+31`, we're overspecifying the test.

The following line (`t.is(getRemoteForLocal(remoteID, 'lo11'), 'ro+20');`) is weak attempt to atone for this sin by adding more information about what we really care about (made even weaker by the fact that the `o+11` comment before it needs to be updated to `o+31`, to match the previous assertion).

A better form of atonement would be to assert the full trio of expected identifiers, starting with the `ro-20` that we made up to represent Alice:

```js
t.is(getLocalForRemote(remoteID, 'ro-20'), 'lo11');
t.is(getKernelForLocal('lo11'), 'o+31');
t.is(getLocalForKernel('o+31'), 'lo11');
t.is(getRemoteForLocal(remoteID, 'lo11'), 'ro+20'); // remember the flip
```

It's basically saying "I know I'm unfairly assuming something about comms' allocation algorithm by asserting `o+31` when really I have no business relying on that choice, but at least here's the full set of unfair assumptions I'm making: I think the `ro-20` I sent in from the remote should map to local `lo11`, and that should map to kernel `o+31`, and back again".

Someone who changes how comms allocates things will see this test fail. They might look at the full set of assertions and realize our overspecification sin, and change the subset of the identifiers necessary to update the sin.

A less sinful way to assert this might be to peek into the comms vat with `getXForY` first. After doing the `d.deliver`, we'd explicitly ask comms what our made-up `ro-20` "Alice" got mapped to, and assert that it shows up in the `syscall.send`:

```js
d.deliver(receiverID, 'receive', ...)
const kernelAlice = getKernelForLocal(getLocalForRemote(remoteID, 'ro-20'));
t.deepEqual(sends.shift(), ... [kernelAlice, bobKernel]) ..
```

I think I didn't do it that way to begin with because it's a bit circular: if e.g. `getKernelForLocal` had a bug where it's allocating `o-NN` instead of `o+NN`, that test wouldn't catch it. It might be useful to assert a tiny bit more about what we get: `t.is(parseVatSlot(kernelAlice).allocatedByVat, true)` or something.

Anyways, basically I took a shortcut when I originally wrote this test, and I feel bad about it, and I figured an extra assertion might make up for it somewhat. But that shouldn't have to be your task. Just update that one `o+11` comment to the new `o+31` and we can do the cleanup some other time.

Reply:
At this point I think we're better off having a more brittle test and then fixing the test when it breaks than we are having a test that uses comms vat code to compute things in a circular fashion, so I think I'll update the comment and add the triumvirate of tests.